nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

```diff
@@ -105,7 +105,7 @@ def write_entry(profile, generation, machine_id):
 
     with open("%s/kernel-params" % (generation_dir)) as params_file:
         kernel_params = kernel_params + params_file.read()
-    with open(tmp_path, 'w') as f:
+    with open(tmp_path, 'wb') as f:
```

Comment:
You will also need to change all write calls below to do bytes but I don't see how this solves encoding issues because it would just shift encoding issues to the string-to-byte conversion step.

Reply:
Yes, this has to be changed everywhere. Is there better way to do this than prefixing every string with `b`?
I don't see any place where we actually have to use strings..

Reply:
For me this does not seem to work:
```python
tmp_path = '/home/user/test"conf.tmp' # working (also without binary mode)
tmp_path = '/boot/loader/entries/test"conf.tmp' # failing

with open(tmp_path, 'wb') as f:
    print("Working")

import os
os.remove(tmp_path)
```
on the vfat `/boot`-partition still throws
```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    with open(tmp_path, 'wb') as f:
OSError: [Errno 22] Invalid argument: '/boot/loader/entries/test"conf.tmp'
```
while the same file on my ext4 Home partition is fine.
Note that I am not even atempting to write - the opening/creating is the problem.

I wonder if the easiest "fix" would be to wrap it into a try/catch block, since I couldn't figure out any solution to this problem and I don't know if limiting the allowed generation-name-chars only because of a systemd-boot entry on an antique file system is a valid reason - #9800 asked for valid chars in general but it is still open

Reply:
It could be that this would resolve the undisplayed non-ASCII characters in the boot menu, though :thinking: But I don't know if the menu supports them at all...

Reply:
Ok, I had hoped that binary mode would fix it. But then maybe its not related to something python does after all.
Could we just name it differently, e.g. encode the profile name by escaping the relevant characters? Apparently https://en.wikipedia.org/wiki/Filename has a list of reserved characters.
Alternatively we could put a tight restriction on profile names instead (something like "alphanumeric").

Reply:
and by escaping I mean encoding really..

Reply:
Ah the file name is never used for actually displaying the entry (as per the [spec](https://systemd.io/BOOT_LOADER_SPECIFICATION/)), right?
So we can just use any encoding we like really, something like [this](https://github.com/acg/python-percentcoding) maybe.

Reply:
You are right - the name is not used for display.

Something I didn't quite get in the spec was, that it said
> For other systems [...] $BOOT must be a VFAT (16 or 32) file system. Applications accessing $BOOT should hence **not assume** that fancier file system features such as symlinks, hardlinks, access control or **case sensitivity** are supported.

and a couple of lines later:
> In order to maximize compatibility [...], file names shall be chosen from a restricted character set: ASCII **upper and lower case characters**, digits, “+”, “-“, “_” and “.”. 

Why does the spec allow upper AND lowercase letters in the filename, if the FS is assumed to be case insensitive? I know far too little about filesystems to answer this, but think that we also currently have problems with having two profiles `test` and `Test` simultaneously. And I don't think we can resolve this if we include the "real" profile name as the differentiating feature in the filename (or we limit it drastically).

My first thought when reading the spec was, to take `basename $(readlink $profile)` (so the hash part of the `/nix/store/` path) as a uniq, non-colliding filename. Including the store-hash would minimize the risk of colisions (e.g. currently if there are two parallel NixOS installs on one machine). But that also doesn't quite work: two profiles can be symlinked to the same actual generation. But I wonder if we could just hash the profile-name (maybe only if not `system`) and append it to the store-path (hashing for keeping the profile-name completely free of limitations)?

Reply:
As long as we can somehow deal with it, I would think it is best to hide the ugly implementation details of a ~35 year old filesystem that many users probably never actively interact with. If there are deeper reasons than the systemd-boot entries, I would totally agree to limit the characterset, but otherwise I think unicode is great (especially for an international audience) :)